src/taipy/core/notification/notifier.py

```diff
@@ -0,0 +1,46 @@
+# Copyright 2023 Avaiga Private Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+# the License. You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+
+from queue import Queue
+from typing import List, Tuple
+
+from .event import Event
+from .registration import Registration
+from .topic import Topic
+
+
+class Notifier:
+    # TODO add a lock on _registration
+    _registrations: List[Registration] = []
+
+    @classmethod
+    def register(cls, entity_type, entity_id, operation, attribute_name, time_to_live) -> Tuple[str, Queue]:
+        register_id = "gui"  # TODO generate an id to return so the client can unregister.
+        registration = Registration(register_id, entity_type, entity_id, operation, attribute_name, time_to_live)
+        # TODO lock before updating cls._registrations
+        cls._registrations.append(registration)
+        return register_id, registration.queue
+
+    @classmethod
+    def unregister(cls, register_id: str = "gui"):
+        # TODO lock before updating cls._registrations
+        cls._registrations = [reg for reg in cls._registrations if reg.register_id != register_id]
+
+    @classmethod
+    def publish(cls, event: Event):
+        for registration in cls._registrations:
+            if cls.is_matching(event, registration.topic):
```

Comment:
Should we add a lock when publishing a new event? As multiple threads would be looking at the queue at this point, we don't have to lock the registration but I think we should lock individual queue when we want to update it. (individual queue instead of all to avoid performance downgrade). But I'm not sure if we have 1 thread putting new element and another getting the element, would that create a conflict? If NO, then locking isn't necessary then?

Reply:
apparently, I think Python has already provided us with a thread-safe queue so we don't need to use any locking mechanism to prevent race condition. Further reading: https://superfastpython.com/thread-queue/

Reply:
If I understand correctly, then we shouldn't need any lock in the orchestrator as well and it should work fine. But maybe we still need it when we want to queue a list of jobs, just to make sure all of the queuing is done prior to starting any executions

Reply:
In the orchestrator, we are using a lock for multiple variables. A queue for the jobs_to_run and a list for the blocked_jobs. So I believe the case is different.

Reply:
Ah you're right :D